xdsl/dialects/tensor.py

```diff
@@ -0,0 +1,35 @@
+from __future__ import annotations
+
+from collections.abc import Sequence
+
+from xdsl.dialects.builtin import IndexType, TensorType
+from xdsl.ir import Attribute, Dialect, SSAValue
+from xdsl.irdl import IRDLOperation, irdl_op_definition, result_def, var_operand_def
+
+
+@irdl_op_definition
+class EmptyOp(IRDLOperation):
+    name = "tensor.empty"
+
+    dynamic_sizes = var_operand_def(IndexType)
+
+    tensor = result_def(TensorType[Attribute])
+
+    assembly_format = (
```

Comment:
So this code was a bit of a hack, that's not quite valid, unfortunately. Here's the link to the op definition in MLIR:

https://mlir.llvm.org/docs/Dialects/TensorOps/#tensorempty-tensoremptyop

The types of the dynamic sizes are actually inside the brackets also I don't know how MLIR handles this in its assembly format (CC @PapyChacal @math-fehr) but, for the time being, we might have to go with a custom parser/printer pair

Reply:
okay, after reading the docs, the only issues i see is with the assembly format and perhaps also needing a check for static and dynamic sizing (i'm not sure if there a functions in xdsl that deal with that perhaps or if it is that important as of rn), other than that i don't see anything else wrong right?

Reply:
i can't also seem to find how mlir handles this too, also what would be the issue with not using the assembly custom format, would that provide issues in parsing and printing this or?

Reply:
that's right, it's just about the dynamic sizes not being empty, and this format definition not being compatible with MLIR's. We probably want to add another test case with a couple of variables in the dynamic sizes variadic operand to check that we can print and parse in the same way as MLIR

Reply:
i've been playing around locally to see if tensor.empty can take the input, but it doesn't like it, as it return "expected 0 operand types but had x", not sure where to go from here, still can't find nothing on how mlir handles this too

Reply:
This is an error from xDSL or MLIR? I would first create an example that works with MLIR, and then implement whatever is needed to make it work in xDSL. I have a feeling that a custom print and parse function will be necessary.

Reply:
this is an error within .mlir, so yh mlir, here is a snippet:
```
%t1 = "tensor.empty"() : () -> tensor<2x3xf32>
%i0 = "test.op"() : () -> index
%t2 = tensor.empty(%i0) : () -> tensor<2xf32>

// CHECK:       module {
// CHECK-NEXT:  %0 = tensor.empty() : tensor<2x3xf32>
// CHECK-NEXT   %1 = test.op() : () -> index
// CHECK-NEXT:  %1 = tensor.empty(%i0) : tensor<2xf32>
// CHECK-NEXT:  }
```
obviously xdsl can take the input but it seems not recognised in mlir

Reply:
Two things:

- MLIR is both the framework and the format. Usually the format manifests itself in textual form in `.mlir` files.
- Both xDSL and MLIR (`xdsl-opt` and `mlir-opt`) are MLIR frameworks, it's just that xDSL is a "deep copy" of the MLIR framework

So, are you getting this error with `mlir-opt` or `xdsl-opt`?

I think `mlir-opt` is right because your operation signature here is different from how you're using it.

```mlir
%t2 = tensor.empty(%i0) : () -> tensor<2xf32>
```

this bit `() -> tensor<2xf32>` says it accepts no args and returns a `tensor<2xf32>`, but you seem to actually pass argument `i0` there.

Reply:
also is that filecheck mixing generic and custom formats for a reason? I didn't check the whole discussion above very closely.

Reply:
oh okay, that makes sense, the error is with ```xdsl-opt```